src/cleanup.ml

```diff
@@ -0,0 +1,180 @@
+                 Xen_api.VBD.plug ~rpc ~session_id ~self:vbd >>= fun () ->
+                 Lwt.finalize
+                   (fun () -> f vbd)
+                   (fun () ->
+                      Lwt_log.notice_f "Unplugging VBD %s" vbd >>= fun () ->
+                      Xen_api.VBD.unplug ~rpc ~session_id ~self:vbd)
+              )
+              (fun () ->
+                 Lwt_log.notice_f "Destroying VBD %s" vbd >>= fun () ->
+                 Xen_api.VBD.destroy ~rpc ~session_id ~self:vbd
+              )
+          )
+      )
+end
+
+module Block = struct
+  module Runtime = struct
+    let blocks_to_close = Hashtbl.create 1
+    let blocks_to_close_mutex = Lwt_mutex.create ()
+
+    let with_tracking b f =
+      let block_uuid = Uuidm.v `V4 |> Uuidm.to_string in
+      Lwt_mutex.with_lock blocks_to_close_mutex (fun () ->
+          Hashtbl.add blocks_to_close block_uuid b; Lwt.return_unit)
+      >>= fun () ->
+      Lwt.finalize
+        f
+        (fun () ->
+           Lwt_mutex.with_lock blocks_to_close_mutex (fun () ->
+               Hashtbl.remove blocks_to_close block_uuid; Lwt.return_unit)
+        )
+
+    let cleanup () =
+      let cleanup b = ignore_exn_log_error "Caught exception while closing open block device" (fun () ->
+          Lwt_log.warning_f "Disconnecting from block device" >>= fun () ->
+          Block.disconnect b)
+      in
+      let blocks_to_close = Hashtbl.fold (fun _ b l -> b::l) blocks_to_close [] in
+      Lwt_list.iter_s cleanup blocks_to_close
+  end
+
+  let with_block filename f =
+    Block.connect filename
+    >>= function
+    | `Error e -> Lwt.fail_with (Printf.sprintf "Unable to read %s: %s" filename (Nbd.Block_error_printer.to_string e))
+    | `Ok b ->
+      Runtime.with_tracking b (fun () ->
+          Lwt.finalize
+            (fun () -> f b)
+            (fun () -> Block.disconnect b)
+        )
+end
+
+module Runtime = struct
+  let cleanup_resources signal =
+    let cleanup () =
+      Lwt_log.warning_f "Caught signal %d, cleaning up" signal >>= fun () ->
+      (* First we have to close the open file descriptors corresponding to the
+         VDIs we plugged to dom0. Otherwise the VDI.unplug call would hang. *)
+      ignore_exn_log_error "Caught exception while closing open block devices" (fun () ->
+          Block.Runtime.cleanup ())
+      >>= fun () ->
+      ignore_exn_log_error "Caught exception while cleaning up VBDs" (fun () ->
+          local_login () >>= fun (rpc, session_id) ->
+          VBD.Runtime.cleanup rpc session_id
+        )
+    in
+
+    Lwt_main.run (cleanup ());
+    failwith (Printf.sprintf "Caught signal %d" signal)
+
+  let register_signal_handler () =
+    let signals = [ Sys.sigint; Sys.sigterm ] in
+    List.iter
+      (fun s -> Lwt_unix.on_signal s cleanup_resources |> ignore)
+      signals
+end
+
+module Persistent = struct
+  let cleanup () =
+    local_login () >>= fun (rpc, session_id) ->
```

Comment:
One thing that I'm not sure about is whether I should log out of these local sessions.

Reply:
I think it better to log out

Reply:
Will do in the the PR.